### PR TITLE
refactor: Introduce ruff and other linting tools

### PR DESCRIPTION
**Pull request by robinholzi**
# Motivation

Previously, we always had to run the `python_compliance.sh`  script, which had intrinsic conflicts so that black and other linters changed each other's reformatting.

I sort was also not even executed on relevant directories like `modern`, leaving us with only sparsely formatted code.

Let's migrate to a ruff, a blazingly fast modern Python linter, and use both ruff linting and formatting on the whole codebase (including directories like `experiments`)!

CI workflows were adjusted accordingly.

### CODE REVIEW

**Comment by robinholzi** [comment]
** Currently, GitHub Actions faces a partial outage; the CI workflow probably still needs some adjustments, so let's wait until https://www.githubstatus.com/ is green again!

File: .github/workflows/workflow.yaml
Comment (MaxiBoether) [inline .github/workflows/workflow.yaml:21]: Any reason for not using the mamba/action.yml that we use for the other jobs?

File: .github/workflows/workflow.yaml
Comment (MaxiBoether) [inline .github/workflows/workflow.yaml:45]: Similarily here, we used to hide the setup of micromamba in the mamba action, not in the task itself. Any reason for changing the mypy job?

File: .github/workflows/workflow.yaml
Comment (MaxiBoether) [inline .github/workflows/workflow.yaml:61]: just to be sure, ruff replaces isort as well?

File: modyn/storage/src/internal/database/sql/SQLUpdateDataset.sql
Comment (MaxiBoether) [inline modyn/storage/src/internal/database/sql/SQLUpdateDataset.sql:1]: Ah, I don't want to vouch for this working. Maybe we want to exclude .sql files from removing whitespaces. if newlines are not equal to a white space here, this would mean `UPDATE datasetsSET base_path` which would be problematic.

File: scripts/python_compliance.sh
Comment (MaxiBoether) [inline scripts/python_compliance.sh:20]: We should document somewhere how to run autoformatting now (and potentially, run it when running this script as before, unless you think something speaks against it)

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Generally LGTM! Thanks so much for taking care of this. 2 notes:

1) Of course before merging we should figure out why CI is consistently failing.

2) Could we document somewhere how to run the formatting? IIRC, the pre commit hook directly modifies my commit inline, whenever I `git commit` right? Because I have faced more than once that autoformatting destroyed something. Hence, I prefer to run autoformatting in a new commit. The pre commit checks action job now covers several checks (ruff, nbstripout, ...), correct?

File: .github/workflows/workflow.yaml
Comment (robinholzi) [reply .github/workflows/workflow.yaml:21]: The local action does some things we don't need for this use case.

File: .github/workflows/workflow.yaml
Comment (robinholzi) [reply .github/workflows/workflow.yaml:45]: If it's just the micromamba setup and nothing else, the indirection through a local actions doesn't really benefit us.

File: .github/workflows/workflow.yaml
Comment (robinholzi) [reply .github/workflows/workflow.yaml:61]: Yes, https://docs.astral.sh/ruff/rules/#isort-i

File: modyn/storage/src/internal/database/sql/SQLUpdateDataset.sql
Comment (robinholzi) [reply modyn/storage/src/internal/database/sql/SQLUpdateDataset.sql:1]: Right, we are using `.sql` files in the most unintuitive way through C++ preprocessor imports😅 In April I spent a couple of hours searching for a better way, but didn't find anything that C++ would accept. Seemingly, you cannot load files in compile time / constexpr functions😅

Comment (robinholzi) [reply modyn/storage/src/internal/database/sql/SQLUpdateDataset.sql:1]: Excluding directories is possible :)

File: scripts/python_compliance.sh
Comment (robinholzi) [reply scripts/python_compliance.sh:20]: autoformatting is done automatically through the pre-commit hooks. You should stage your changes before running a git commit to be able to undo things. If you don't want the pre-commit hooks to run on every commit, you should alias your git commit command with a `-n` flag. That's nothing that should be configured centrally.

Comment (robinholzi) [reply scripts/python_compliance.sh:20]: I didn't want to include pylint in the pre-commit hooks, as it's prohibitively slow (20-30 sec.). The other hooks run in <5s total. It's still in the python_compliance script though (for CI)

File: modyn/storage/src/internal/database/sql/SQLUpdateDataset.sql
Comment (robinholzi) [reply modyn/storage/src/internal/database/sql/SQLUpdateDataset.sql:1]: Doesn't seem to be harmful though

File: .github/workflows/workflow.yaml
Comment (MaxiBoether) [reply .github/workflows/workflow.yaml:45]: But I mean the local action takes care of caching, no? In case of no updates, we can restore the environment from cache, speeding up the CI - that is what the local action does

File: .github/workflows/workflow.yaml
Comment (MaxiBoether) [reply .github/workflows/workflow.yaml:21]: Similar to my other comment, it caches the environment though, doesn't it? And that speeds up the action

**Comment by MaxiBoether** [comment]
We can merge this after figuring out the setup action. The reason I suggest to use that one is that it caches the environment.

**Review by MaxiBoether — APPROVED** [review]
Giving my approval here but let's figure out the caching thing

**Comment by robinholzi** [comment]
> Giving my approval here but let's figure out the caching thing

I think we don't need this fancy caching setup with date and cache version. I'm removing this and defaulting to the default cache environment across all setup-micromamba occurrences. Let me you if there's a strong reason to have this, than I can restore it in a follow-up PR.

File: .github/workflows/workflow.yaml
Comment (robinholzi) [reply .github/workflows/workflow.yaml:45]: we already did that with the `cache-environment` micromamba flag, just with a different (default) cache-key. Now we use the default cache environment everywhere.

File: .github/workflows/workflow.yaml
Comment (robinholzi) [reply .github/workflows/workflow.yaml:21]: The action runs in <30 sec. We don't need to optimize this further.

**Pull request merged**
